lib/r10k/api/modules.rb

```diff
@@ -195,6 +202,53 @@ def self.resolve_forge_module(mod, opts={})
       end
       private_class_method :resolve_forge_module
 
+      # Cache a specific module release tarball from the Puppet Forge.
+      #
+      # @param release_slug [String] Hyphen separated namespace, module name, and release of the module to be cached or updated. (E.g. "puppetlabs-apache")
+      # @param force [Boolean] Whether or not to overwrite an existing release tarball if present.
+      # @option opts [String] :cachedir Base path where caches are stored.
+      # @option opts [Hash] :forge Additional options to control interaction with a Puppet Forge API implementation.
+      # @return [String] Returns the full path to the tarball on success, raises on failure.
+      # @raise [RuntimeError]
+      # @api private
+      def self.get_cached_forge_release(release_slug, cachedir, force = false, forge_opts = {})
+        module_slug = module_slug_from_release_slug(release_slug)
+        tarball_cachedir = cachedir_for_forge_module(module_slug, cachedir)
+
+        # TODO: Implement forge options, probably with supporting code in puppet_forge gem.
+        if !forge_opts.empty?
+          raise NotImplementedError "Updating the forge cached modules does not support forge options: #{forge_opts}"
+        end
+
+        if !File.directory?(tarball_cachedir)
+          File.mkdir_p(tarball_cachedir)
+        end
+
+        final_destination = File.join(tarball_cachedir, release_slug + ".tar.gz")
+
+        if !File.exists?(final_destination) || force
+
+          Dir.mktmpdir(release_slug, tarball_cachedir) do |tmpdir|
+            tmp_destination = File.join(tmpdir, release_slug + ".tar.gz")
+
+            begin
+              PuppetForge::Release.find(release_slug).tap do |release|
+                release.download(Pathname(tmp_destination))
+                release.verify(Pathname(tmp_destination))
+              end
+            rescue PuppetForge::ReleaseNotFound, PuppetForge::ReleaseForbidden => e
+              raise RuntimeError.new e.message
```

Comment:
This should include the backtrace from the original error; otherwise this will generate a new backtrace and may hide where the error actually came from. https://github.com/puppetlabs/r10k/blob/master/lib/r10k/errors.rb#L10-L22 implements exception wrapping which might come in handy here.